utils/utils_quantization.py

```diff
@@ -0,0 +1,312 @@
+import gudhi as gd
+
+
+def _fparab(a,b,x):
+    return (x - a)**2 / (2 * b) + (b / 2)
+
+
+def _rotate(t):
+    res = np.dot(t, np.sqrt(2)/2 * np.array([[1, 1],[-1, 1]]))
+    return res
+
+
+def _parab(c):
+    a, b = c
+    def P(x):
+        return _fparab(a,b,x) # (x**2 - 2 * a * x + b**2 + a**2)/(2 * b)
+    return P
+
+
+def plot_partition_codebook(cs):
+    xmin, xmax = -7, 3
+    ymin, ymax = -2, 8
+
+    x = np.linspace(-10, 10, 100)
+    ys = np.array([_parab(c)(x) for c in cs])
+    miny = np.min(ys, axis=0)
+
+    r_cs = _rotate(cs)
+
+    vor = Voronoi(r_cs)
+
+    fig, ax = plt.subplots(figsize=(6,6))
+
+    voronoi_plot_2d(vor, ax, show_points=False, show_vertices=False)
+    ax.scatter(r_cs[:,0], r_cs[:,1], marker='o', c='b')
+
+    ax.annotate('$c_j$', r_cs[0]+[0.2,0.2], c='blue', fontsize=24)
+    #ax.annotate('$c_2$', r_cs[2]+[0.2,0.2], c='green', fontsize=24)
+    #ax.annotate('$V_j(\mathbf{c})$', r_cs[1]+[-0.2,1.9], c='blue', fontsize=24)
+    #ax.annotate('$V_{k+1}(\mathbf{c})$', r_cs[1]+[2.,-2], c='black', fontsize=24, rotation=45)
+
+    tmp = np.zeros((len(x), 2))
+    tmp[:,0] = x
+    tmp[:,1] = miny
+    r_parab = _rotate(tmp)
+    ax.plot(r_parab[:,0], r_parab[:,1], linestyle='dashed', color='black', linewidth=3)
+
+    ax.set_aspect('equal')
+    ax.fill_between(r_parab[:60,0],y1=r_parab[:60,0],y2=r_parab[:60,1], color='white', alpha=1, zorder=3)
+    ax.fill_betweenx(r_parab[59:,1],x1=r_parab[59:,1],x2=r_parab[59:,0], color='white', alpha=1, zorder=3)
+    ax.add_patch(mpatches.Polygon([[0,0], [0,r_parab[59,1]], [r_parab[59,1], r_parab[59,1]]], fill=True, color='white', alpha=1, zorder=3))
+    ax.add_patch(mpatches.Polygon([[ymin,ymin], [xmax,ymin], [xmax, xmax]], fill=True, color='lightgrey', alpha=1,zorder=3))
+    ax.plot([min(xmin,ymin), max(xmax,ymax)], [min(xmin,ymin), max(xmax,ymax)], color='k', linewidth=3,zorder=3)
+    ax.annotate('$\partial \Omega$', [-1.3, -1.9], fontsize=24)
+    #ax.annotate('$N(\mathbf{c})$', [-6.5,0], fontsize=24, rotation=45)
+
+    ax.set_axis_off()
+    ax.set_ylim(ymin,ymax)
+    ax.set_xlim(xmin, xmax)
+
+
+def plot_dgm(dgm, box=None, ax=None, color="blue", label='diagram', alpha=None):
+    if ax is None:
+        fig = plt.figure()
+        ax = fig.add_subplot(111)
+    if dgm.size:
+        x, y = dgm[:, 0], dgm[:, 1]
+        ax.scatter(x, y, marker='o', color=color, label=label, alpha=alpha)
+    if box is None:
+        if dgm.size:
+            m, M = np.min(dgm) - 0.1, np.max(dgm) + 0.1
+        else:
+            m, M = 0, 1
+    else:
+        m, M = box
+    ax.set_xlim(m, M)
+    ax.set_ylim(m, M)
+    ax.add_patch(mpatches.Polygon([[m, m], [M, m], [M, M]], fill=True, color='lightgrey'))
+    ax.set_aspect("equal")
+    ax.set_xlabel("Birth", fontsize=24)
+    ax.set_ylabel("Death", fontsize=24)
+    # ax.legend()
+
+
+########################
+### Experiment utils ###
+########################
+
+
+def _sample_torus(n, r1, r2, radius_eps, ambiant_eps):
```

Comment:
Is it the same as in the other utils file?
Ah, I think I understand, each notebook imports only one file. If github respected the tag to hide/remove/collapse a cell, this would all be in the notebook, but since it doesn't, you put it in a separate file. If that's the case (and it makes sense, in some sense I prefer that instead of developing a library bis in the utils of the notebooks), I'd like the separate file to have a name that clearly matches the name of the corresponding notebook.

Reply:
Yes, my idea was "ok I don't want to overload a given notebook with plot function & cie", so I put such functions in a specific `utils_[ref-to--the-notebook].py` file. It possibly allows to fine-tune the functions for each specific notebook : e.g. I may want some day to enhance the `plot_diag` for a given notebook for some reason without changing the one of another notebook.